src/rgw/driver/rados/rgw_data_sync.cc

```diff
@@ -5614,10 +5619,32 @@ int RGWSyncBucketCR::operate(const DoutPrefixProvider *dpp)
         // lease not required for incremental sync
         RELEASE_LOCK(bucket_lease_cr);
 
+        assert(sync_pair.source_bs.shard_id >= 0);
         // if a specific gen was requested, compare that to the sync status
         if (gen) {
-          const auto current_gen = bucket_status.incremental_gen;
+          current_gen = bucket_status.incremental_gen;
+          source_bs = sync_pair.source_bs;
           if (*gen > current_gen) {
+            /* In case the data log entry is missing for previous gen, it may
+             * not be marked complete and the sync can get stuck. To avoid it,
+             * may be we can add this (shardid, gen) to error repo to force
+             * sync and mark that shard as completed.
+             */
+            pool = sc->env->svc->zone->get_zone_params().log_pool;
```

Comment:
this seems like an interesting approach, and doesn't seem to cost anything if we don't see the potentially stale generation

Reply:
Thats right!